internal/server/routes.go

```diff
@@ -297,6 +304,33 @@ type isBlockingRequest interface {
 	IsBlockingRequest() bool
 }
 
+func requestVersion(req *http.Request) (*semver.Version, error) {
+	headerVer := req.Header.Get("Infra-Version")
+	if headerVer == "" {
+		return nil, fmt.Errorf("%w: Infra-Version header is required. The current version is %s", internal.ErrBadRequest, internal.FullVersion())
+	}
+	reqVer, err := semver.NewVersion(headerVer)
+	if err != nil {
+		return nil, fmt.Errorf("%w: invalid Infra-Version header: %v. Current version is %s", internal.ErrBadRequest, err, internal.FullVersion())
+	}
+	return reqVer, nil
+}
+
+func handlerForVersion(versions []routeVersion, reqVer *semver.Version) func(c *gin.Context) {
+	if reqVer == nil {
+		return nil
+	}
+
+	// TODO: iterate in reverse order to benefit newer versions over older versions
+	for _, v := range versions {
+		if reqVer.GreaterThan(v.version) {
+			continue
+		}
+		return v.handler
+	}
+	return nil
+}
```

Comment:
This function is really the only new code in this PR. These 13 lines replace ~400 lines of migration code.

Reply:
This looks good, but as far as I can see there's nowhere that grantees the `versions` slice is in the correct order (oldest version to newest version). Should we sort it before applying the migrations?

Reply:
Yes, great point! That is something I forgot to add. 

We could do that at runtime before returning the routes, but I'm tempted to do it via a test (which will fail if any routes are in the wrong order). That way the routes should always be listed in the correct order in the source code.